src/spi/blocking.rs

```diff
@@ -1,108 +1,338 @@
+//! pub struct MyDriver<SPI> {
+//!     spi: SPI,
+//! }
+//!
+//! impl<SPI> MyDriver<SPI>
+//! where
+//!     SPI: SpiDevice,
+//!     SPI::Bus: SpiBus, // or SpiBusRead/SpiBusWrite if you only need to read or only write.
+//! {
+//!     pub fn new(spi: SPI) -> Self {
+//!         Self { spi }
+//!     }
+//!
+//!     pub fn read_foo(&mut self) -> [u8; 2] {
+//!         let mut buf = [0; 2];
+//!
+//!         // `transaction` asserts and deasserts CS for us. No need to do it manually!
+//!         self.spi.transaction(|bus| {
+//!             bus.write(&[0x90]).unwrap();
+//!             bus.read(&mut buf).unwrap();
+//!         }).unwrap();
+//!
+//!         buf
+//!     }
+//! }
+//! ```
+//!
+//! If your device **does not have a CS pin**, use [`SpiBus`] (or [`SpiBusRead`], [`SpiBusWrite`]). This will ensure
+//! your driver has exclusive access to the bus, so no other drivers can interfere. It's not possible to safely share
+//! a bus without CS pins. By requiring [`SpiBus`] you disallow sharing, ensuring correct operation.
+//!
+//! ```
+//! # use embedded_hal::spi::blocking::{SpiBus, SpiBusRead, SpiBusWrite};
+//! pub struct MyDriver<SPI> {
+//!     spi: SPI,
+//! }
+//!
+//! impl<SPI> MyDriver<SPI>
+//! where
+//!     SPI: SpiBus, // or SpiBusRead/SpiBusWrite if you only need to read or only write.
+//! {
+//!     pub fn new(spi: SPI) -> Self {
+//!         Self { spi }
+//!     }
+//!
+//!     pub fn read_foo(&mut self) -> [u8; 2] {
+//!         let mut buf = [0; 2];
+//!         self.spi.write(&[0x90]).unwrap();
+//!         self.spi.read(&mut buf).unwrap();
+//!         buf
+//!     }
+//! }
+//! ```
+//!
+//! If you're (ab)using SPI to **implement other protocols** by bitbanging (WS2812B, onewire, generating arbitrary waveforms...), use [`SpiBus`].
+//! SPI bus sharing doesn't make sense at all in this case. By requiring [`SpiBus`] you disallow sharing, ensuring correct operation.
+//!
+//! # For HAL authors
+//!
+//! HALs **must** implement [`SpiBus`], [`SpiBusRead`] and [`SpiBusWrite`]. Users can combine the bus together with the CS pin (which should
+//! implement [`OutputPin`]) using HAL-independent [`SpiDevice`] implementations such as [`ExclusiveDevice`].
+//!
+//! HALs may additionally implement [`SpiDevice`] to **take advantage of hardware CS management**, which may provide some performance
+//! benefits. (There's no point in a HAL implementing [`SpiDevice`] if the CS management is software-only, this task is better left to
+//! the HAL-independent implementations).
+//!
+//! HALs **must not** add infrastructure for sharing at the [`SpiBus`] level. User code owning a [`SpiBus`] must have the guarantee
+//! of exclusive access.
+//!
+//! # Flushing
+//!
+//! To improve performance, Bus implementations are allowed to return before the operation is finished, i.e. when the bus is still not
+//! idle. You must call [`flush`](SpiBusFlush::flush) to wait for operations to actually finish, for example before deasserting CS in a
```

Comment:
i'm not sure exactly how to reframe this but, it might be nice to separate the `Bus` and `Device` components, something like:

```suggestion
//! idle. When using a `Bus` directly if you need to ensure the bus is idle you can manually call [`flush`](SpiBusFlush::flush).
//!  [`SpiDevice`] implementations **must** call [`flush`](SpiBusFlush::flush) prior to de-asserting CS to maintain ordering guarantees.
```

Reply:
You may still have to flush even if using a SpiDevice, for example in an SPI display to ensure the DCX pin toggles at the right time wrt the spi data. It's rare though....

Reply:
```
, while [`SpiDevice`] users may choose to flush within an transaction if required.
```

Reply:
I've expanded docs on when flush might be needed, hopefully it's much straightforward now!